airsonic-main/src/main/java/org/airsonic/player/service/StatusService.java

```diff
@@ -43,35 +45,29 @@
     @Autowired
     private MediaFileService mediaFileService;
 
-    private final List<TransferStatus> streamStatuses = new ArrayList<TransferStatus>();
-    private final List<TransferStatus> downloadStatuses = new ArrayList<TransferStatus>();
-    private final List<TransferStatus> uploadStatuses = new ArrayList<TransferStatus>();
-    private final List<PlayStatus> remotePlays = new ArrayList<PlayStatus>();
+    private final List<TransferStatus> streamStatuses = Collections.synchronizedList(new ArrayList<>());
+    private final List<TransferStatus> downloadStatuses = Collections.synchronizedList(new ArrayList<>());
+    private final List<TransferStatus> uploadStatuses = Collections.synchronizedList(new ArrayList<>());
+    private final List<PlayStatus> remotePlays = Collections.synchronizedList(new ArrayList<>());
 
     // Maps from player ID to latest inactive stream status.
-    private final Map<Integer, TransferStatus> inactiveStreamStatuses = new LinkedHashMap<Integer, TransferStatus>();
-
-    public synchronized TransferStatus createStreamStatus(Player player) {
-        // Reuse existing status, if possible.
-        TransferStatus status = inactiveStreamStatuses.get(player.getId());
-        if (status != null) {
-            status.setActive(true);
-        } else {
-            status = createStatus(player, streamStatuses);
-        }
-        return status;
+    private final Map<Integer, TransferStatus> inactiveStreamStatuses = new ConcurrentHashMap<>();
+
+    public TransferStatus createStreamStatus(Player player) {
+        return createStatus(player, streamStatuses);
     }
 
     public synchronized void removeStreamStatus(TransferStatus status) {
         // Move it to the map of inactive statuses.
-        status.setActive(false);
-        inactiveStreamStatuses.put(status.getPlayer().getId(), status);
-        streamStatuses.remove(status);
+        inactiveStreamStatuses.compute(status.getPlayer().getId(), (k, v) -> {
```

Comment:
Is it normal that neither `k` nor `v` is used? I'm not sure I understand the use of `compute` in that case.

Reply:
Because it's what `compute` is expecting: https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#compute-K-java.util.function.BiFunction-

I copy-pasted this from @randomnicode 's code, but I can remove it, since it doesn't seem super-useful in this case.

Reply:
Both `k` (the key to bind the new value to) and `v` (the old value, if any) are not really needed. The key-binding is implicit, and we discard the old value (if present). We just need to associate the key with the new value.

There is a reason, however, why `compute` is used and that's because of concurrency. `inactiveStreamStatuses` is a `ConcurrentHashMap`. Using `compute` guarantees atomicity with respect to incoming thread requests. In fact, this pull request seems incomplete to me because that is the very reason you can remove the synchronized lock on this entire method, entrusting the ConcurrentHashmap to do its job.

Here's what might happen without the compute AND the sync:
- thread1: removeStreamStatus(statusA)
- thread2: removeStreamStatus(statusB)
- thread1: statusA.setActive()
- thread2: statusB.setActive()
- thread2: inactiveStreamStatuses.put(id, statusB) <- preempted
- thread1:  inactiveStreamStatuses.put(id, statusA)
...

statusA is followed by statusB, you would expect statusB to be the final inactiveStream for that id, but statusA comes out bound at the end.

With syncs, you can achieve good order by locking every request out and making them wait in a serial order.

With concurrent maps, you can do even better by making only some requests wait (ids in the same hash bucket), and letting everything else transact concurrently.

Again, this PR seems incomplete to me. I likely made more modifications later? The current version of this class in master (or even in the pointed commit: https://github.com/airsonic-advanced/airsonic-advanced/commit/caa5f8346cc4597517e25847780a98af0f3432ec#diff-9cf7e6b3f2a25472d4d3e4220a8dcb71R63 ) does not have a sync in the method signature

Reply:
> With concurrent maps, you can do even better by making only some requests wait (ids in the same hash bucket), and letting everything else transact concurrently.

Thanks, that is a good reason.

In any case, swapping an explicit `synchronized` with an implicit "you should know that compute is synchronized on the internal buckets so it's good for performance!" is not ideal.

I'm okay with either the explicit `synchronized` _or_ having a comment above it making sure that someone who reads this understands this.

Reply:
> In any case, swapping an explicit synchronized with an implicit "you should know that compute is synchronized on the internal buckets so it's good for performance!" is not ideal.

It's not implicit, it's part of the `ConcurrentHashMap` contract itself and is written in the javadoc pretty clearly. There are no hidden gotchas here:
https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#compute-K-java.util.function.BiFunction-

Here's the explicit portion:
>Attempts to compute a mapping for the specified key and its current mapped value (or null if there is no current mapping). The entire method invocation is performed atomically. Some attempted update operations on this map by other threads may be blocked while computation is in progress, so the computation should be short and simple, and must not attempt to update any other mappings of this Map.